src/Symfony/Component/Security/Http/Authenticator/AbstractPreAuthenticatedAuthenticator.php

```diff
@@ -0,0 +1,106 @@
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+
+/**
+ * The base authenticator for authenticators to use pre-authenticated
+ * requests (e.g. using certificates).
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @internal
+ * @experimental in Symfony 5.1
+ */
+abstract class AbstractPreAuthenticatedAuthenticator implements InteractiveAuthenticatorInterface
+{
+    private $userProvider;
+    private $logger;
+
+    public function __construct(UserProviderInterface $userProvider, ?LoggerInterface $logger = null)
+    {
+        $this->userProvider = $userProvider;
+        $this->logger = $logger;
+    }
+
+    /**
+     * Returns the username of the pre-authenticated user.
+     *
+     * This authenticator is skipped if null is returned or a custom
+     * BadCredentialsException is thrown.
+     */
+    abstract protected function extractUsername(Request $request): ?string;
+
+    public function supports(Request $request): ?bool
+    {
+        try {
+            $username = $this->extractUsername($request);
+        } catch (BadCredentialsException $e) {
+            if (null !== $this->logger) {
+                $this->logger->debug('Skipping pre-authenticated authenticator as a BadCredentialsException is thrown.', ['exception' => $e, 'authenticator' => static::class]);
+            }
+
+            return false;
+        }
+
+        if (null === $username) {
+            if (null !== $this->logger) {
+                $this->logger->debug('Skipping pre-authenticated authenticator no username could be extracted.', ['authenticator' => static::class]);
+            }
+
+            return false;
+        }
+
+        $request->attributes->set('_pre_authenticated_username', $username);
+
+        return true;
+    }
+
+    public function authenticate(Request $request): PassportInterface
+    {
+        $username = $request->attributes->get('_pre_authenticated_username');
+        $user = $this->userProvider->loadUserByUsername($username);
+
+        return new SelfValidatingPassport($user);
+    }
+
+    public function createAuthenticatedToken(PassportInterface $passport, string $providerKey): TokenInterface
+    {
+        return new PreAuthenticatedToken($passport->getUser(), null, $providerKey, $passport->getUser()->getRoles());
+    }
+
+    public function onAuthenticationSuccess(Request $request, TokenInterface $token, string $providerKey): ?Response
+    {
+        return null; // let the original request continue
+    }
+
+    public function onAuthenticationFailure(Request $request, AuthenticationException $exception): ?Response
+    {
+        return null;
+    }
```

Comment:
In the old `AbstractPreAuthenticatedListener`, on failure, it called `AbstractPreAuthenticatedListener::clearToken()`, which removed the token from the token storage. In this new system, we purposely *stopped* doing this - you even include a description of this in the PR. However, now I see some rationale for it - at least in this case. Consider this situation:

1) I authenticate via `remote_user`. Forgive me, I know almost nothing about `remote_user`, so I could be making some false assumptions.
2) The token gets stored in the session because I'm using a stateful firewall.
3) My `remote_user` (or `x509`) authentication is *revoked*. If we do NOT clear the token from the storage, it may allow the user to be continued to stay authenticated, even though some central system has removed their access.

WDYT? It might be safest to re-add this token clearing everywhere. Or, perhaps different authenticators have different behavior: I'm not sure it makes sense for `form_login` to have this same behavior.

Reply:
So then it would make most sense to do this in `AuthenticatorManager` for all authentication failures.

Reply:
Indeed it may... though I still don't love that happening with, for example, a "login form" system :/